consensus/service/src/api/client_api_service.rs

```diff
@@ -27,15 +27,55 @@ use mc_ledger_db::Ledger;
 use mc_peers::ConsensusValue;
 use mc_transaction_core::mint::{MintConfigTx, MintTx};
 use mc_util_grpc::{check_request_chain_id, rpc_logger, send_result, Authenticator};
-use std::sync::{Arc, Mutex};
+use std::{
+    collections::VecDeque,
+    sync::{Arc, Mutex},
+    time::{Duration, Instant},
+};
 
 /// Maximum number of pending values for consensus service before rejecting
 /// add_transaction requests.
 const PENDING_LIMIT: i64 = 500;
 
 /// Data retained on a session with a client.
 #[derive(Clone, Debug)]
-pub struct ClientSessionTracking {/* stub - todo: track failed TX proposals */}
+pub struct ClientSessionTracking {
+    // This needs to be a VecDeque because popping elements oldest-first (i.e.
+    // in a FIFO fashion) needs to be efficient, since we'll be culling the
+    // oldest tx failure timestamps constantly - if the server is configured to
+    // drop clients with over 100 failed tx proposals in the last 30 seconds,
+    // we don't care about timestamps from over 30 seconds ago, for example.
+    tx_proposal_failures: VecDeque<Instant>,
+}
+
+impl ClientSessionTracking {
+    pub fn new() -> Self {
+        Self {
+            tx_proposal_failures: VecDeque::default(),
+        }
+    }
+
+    /// Push a new failed tx proposal record, clear out samples older than
+    /// our tracking window, and return the number of tx failures remaining
+    /// on the list - as-in, tells you "there have been x number of failures
+    /// within the past `tracking_window` seconds".
+    ///
+    /// # Arguments
+    ///
+    /// * `now` - Used as both the instant to record as a new tx proposal
+    /// failure, and as the "present" time to check for stale records.
+    /// * `tracking_window` - How long of a period of time should we keep track
+    /// of an individual tx proposal failure incident for? Any records which
+    /// have existed for longer than this value will be dropped when this
+    /// method is called.
+    pub fn fail_tx_proposal(&mut self, now: &Instant, tracking_window: &Duration) -> usize {
```

Comment:
Why use an `&Instant` instead of an `Instant` if you dereference it on line 75? I'd recommend taking an owned `Instant`. Same for `&Duration`- I'd recommend using a `Duration`.

Reply:
Addressed in [6b3861f](https://github.com/mobilecoinfoundation/mobilecoin/pull/3208/commits/6b3861fa7beed20555822ea943a9552d3c9d035b)